homeassistant/components/esphome/config_flow.py

```diff
@@ -53,6 +53,23 @@ def __init__(self):
             errors=errors
         )
 
+    async def async_step_discovery(self, user_input: ConfigType):
+        """Handle discovery."""
+        # mDNS hostname has additional '.' at end
+        hostname = user_input['hostname'].replace('.local.', '.local')
+        hosts = (hostname, user_input['host'])
+        for entry in self._async_current_entries():
+            if entry.data['host'] in hosts:
+                return self.async_abort(
+                    reason='already_configured'
+                )
+
+        # Prefer .local addresses (mDNS is available after all, otherwise
+        # we wouldn't have received the discovery message)
+        self._host = hostname
+        self._port = user_input['port']
+        return await self.async_step_user()
```

Comment:
I prefer an `async_step_confirm`, no more data entry.

Reply:
Ended up doing it with 

```yaml
return await self.async_step_user(user_input={
    'host': hostname,
    'port': user_input['port'],
})
```

which also handles connection errors + asking for password.